Kernel/TTY/TTY.h

```diff
@@ -91,7 +91,8 @@ class TTY : public CharacterDevice {
     virtual bool is_tty() const final override { return true; }
 
     CircularDeque<u8, 1024> m_input_buffer;
-    WeakPtr<Process> m_process;
+    WeakPtr<Process> m_original_process_parent;
+    ProcessGroupID m_pgid { 0 };
```

Comment:
But what happens when this PGID dies?

Reply:
Nothing interesting.
we still act as if the pgid existed and `SIGTTIN`/`SIGTTOU` everyone that doesn't have the (now invalid) pgid.

Reply:
So the shell has to "grab back" the tty? This makes a lot of sense, since something else had to "grab" the tty earlier.

Headsup: I don't know how it works on Linux, so maybe you need to detect whether the Shell is running in Lagom or in Serenity.

Reply:
But then this is vulnerable to PID reuse attacks, no?

Reply:
> PID reuse

well, it'd be a PGID reuse "attack", not sure what the use of such a thing would be.
In a sense, the TTY is owned by a PGID, not caring about who _is_ in that process group.

Not sure how we would detect that every process in a pgrp died (and even if that's a sensible thing to do!)

Reply:
> I don't know how it works on Linux

Not sure about other places, but the TTY only tracks the PGID, and somehow detects "orphaned" pgrp's, disallowing writes.

Reply:
> well, it'd be a PGID reuse "attack", not sure what the use of such a thing would be.

I'd rather we don't leave this door open. :)

Reply:
Fixing _that_ specific problem would be a bit more involved - we'd have to figure out when all members of a pgrp die, but I don't think we have an abstraction for process groups, they're just a field on actual processes.

I'll mark this PR as a draft until we figure something out for this case then.

Reply:
One solution would be to have a `ProcessGroup` object in the kernel.

Reply:
One way would be to track *some* Process of the owning PGID, and whenever it dies, grab any other process with the right PGID. That's slow as hell, but process creation/destruction shouldn't happen that often, especially not without a leader.

This would get the PR unstuck, no?

Reply:
Nope, that leaves a window of vulnerability where no one is in a process group:

pgid P: {A,B}
TTY proc: A
(A dies)
(B dies)
...
(C makes pgid P)
pgid P: {C}
(C writes to the TTY)
TTY looks for another member in the PGID and finds C
(TTY allows C to do whatever)